llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp

```diff
@@ -795,6 +795,14 @@ static bool canWidenCallReturnType(Type *Ty) {
 bool LoopVectorizationLegality::canVectorizeInstrs() {
   BasicBlock *Header = TheLoop->getHeader();
 
+  /// Generally, if `TheLoop` lacks a PreHeader, it won't reach this point.
+  /// However, when executed with remarks options like `-pass-remarks=vector`
+  /// it will reach here to output extra debug messages, so need to check.
+  /// Additionally, before reaching here, the debug messages would have
+  /// included `Loop doesn't have a legal pre-header`, so we omit that.
+  if (ORE->allowExtraAnalysis(DEBUG_TYPE) && !TheLoop->getLoopPreheader())
+    return false;
```

Comment:
Checking here means we will miss out on any extra analysis for subsequent instructions.

We should be able to check in IVDescriptor and bail out if there's no preheader there, which would still allow continuing with the other checks.

You can add an unvectorizable instruction to the function, perhaps something like an unsupported call, for which we should get a missed remark ideally, even if there's no preheader.

Reply:
I understand that you are asking me to modify the subcode so that no issues arise even if the PreHeader does not exist. Is that correct?

Reply:
> I understand that you are asking me to modify the subcode so that no issues arise even if the PreHeader does not exist. Is that correct?

Yep.

>>You can add an unvectorizable instruction to the function, perhaps something like an unsupported call, for which we should get a missed remark ideally, even if there's no preheader.

Looks like this is still pending?